File: frontend/src/types/task.ts
export type Priority = 'low' | 'medium' | 'high'
export type Status = 'waiting' | 'in_progress' | 'done'

export interface Task {
  id: string
  title: string
  description: string | null
  priority: Priority
  status: Status
  deadline: string | null
  category: string | null
  created_at: string
  updated_at: string
}

export interface TaskCreate {
  title: string
  description?: string | null
  priority: Priority
  status: Status
  deadline?: string | null
  category?: string | null
}

export interface TaskUpdate {
  title?: string
  description?: string | null
  priority?: Priority
  status?: Status
  deadline?: string | null
  category?: string | null
}

export interface TaskListResponse {
  items: Task[]
  total: number
}

export interface TaskFilters {
  status?: Status | ''
  priority?: Priority | ''
  search?: string
  deadline_before?: string
}

export interface CategorySuggestion {
  category: string
  reasoning: string
}

export interface SubTask {
  title: string
  description?: string | null
}

export interface DecompositionResult {
  subtasks: SubTask[]
  reasoning: string
}

export interface PrioritySuggestion {
  priority: Priority
  reasoning: string
}

export interface WorkloadSummary {
  summary: string
  overdue_count: number
  upcoming_count: number
  distribution: Record<string, number>
  from_cache?: boolean
}

File: backend/app/schemas/task.py
from pydantic import BaseModel, field_validator
from datetime import datetime, date
from typing import Optional
from app.models.task import Priority, Status


class TaskBase(BaseModel):
    title: str
    description: Optional[str] = None
    priority: Priority = Priority.medium
    status: Status = Status.waiting
    deadline: Optional[date] = None
    category: Optional[str] = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Title must not be empty")
        return v.strip()


class TaskCreate(TaskBase):
    pass


class TaskUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[Priority] = None
    status: Optional[Status] = None
    deadline: Optional[date] = None
    category: Optional[str] = None

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("Title must not be empty")
        return v.strip() if v else v


class TaskResponse(TaskBase):
    id: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class TaskListResponse(BaseModel):
    items: list[TaskResponse]
    total: int


class ErrorResponse(BaseModel):
    detail: str
    code: Optional[str] = None
